crates/flightsql_datafusion/src/lib.rs

```diff
@@ -0,0 +1,465 @@
+
+    #[snafu(display("Unable to query FlightSQL: {source}"))]
+    Flight { source: flight_client::Error },
+
+    #[snafu(display("Unable to query FlightSQL: {source}"))]
+    ArrowFlight { source: FlightError },
+
+    #[snafu(display("Unable to retrieve schema"))]
+    NoSchema,
+}
+
+type Result<T, E = FlightSQLError> = std::result::Result<T, E>;
+
+pub struct FlightSQLTable {
+    client: FlightSqlServiceClient<Channel>,
+    table_reference: OwnedTableReference,
+    schema: SchemaRef,
+}
+
+#[allow(clippy::needless_pass_by_value)]
+impl FlightSQLTable {
+    pub async fn new(
+        client: FlightSqlServiceClient<Channel>,
+        table_reference: impl Into<OwnedTableReference>,
+    ) -> Result<Self> {
+        let table_reference: OwnedTableReference = table_reference.into();
+        let schema = Self::get_schema(client.clone(), table_reference.clone()).await?;
+        Ok(Self {
+            client,
+            table_reference,
+            schema,
+        })
+    }
+
+    pub async fn from_static(
+        s: &'static str,
+        table_reference: impl Into<OwnedTableReference>,
+    ) -> Result<Self> {
+        let channel = channel::Endpoint::from_static(s)
+            .connect()
+            .map_err(|_| FlightSQLError::UnableToConnectToServer)
+            .await?;
+        Self::new(FlightSqlServiceClient::new(channel), table_reference.into()).await
+    }
+
+    fn get_str_from_record_batch(b: &RecordBatch, row: usize, col_name: &str) -> Option<String> {
+        if let Some(col_array) = b.column_by_name(col_name) {
+            if let Some(y) = col_array.as_any().downcast_ref::<array::StringArray>() {
+                return Some(y.value(row).to_string());
+            }
+        }
+        None
+    }
+
+    #[must_use]
+    pub fn get_table_schema_if_present(
+        batches: Vec<RecordBatch>,
+        table_reference: OwnedTableReference,
+    ) -> Option<SchemaRef> {
+        let schema_bytz_opt = batches.iter().find_map(|b| {
+            let table_schema = b
+                .column_by_name("table_schema")
+                .and_then(|ts_array| ts_array.as_any().downcast_ref::<array::BinaryArray>())
+                .or(None)?;
+
+            for i in 0..b.num_rows() {
+                let table_name =
+                    Self::get_str_from_record_batch(b, i, "table_name").unwrap_or_default();
+                let catalog_name =
+                    Self::get_str_from_record_batch(b, i, "catalog_name").unwrap_or_default();
+                let db_schema_name =
+                    Self::get_str_from_record_batch(b, i, "db_schema_name").unwrap_or_default();
+
+                // We got at least one None above, so we can't compare.
+                if table_name.is_empty() || catalog_name.is_empty() || db_schema_name.is_empty() {
+                    continue;
+                }
+
+                // Match even if catalog_name or db_schema_name from table_reference was None.
```

Comment:
This code won't return the table_schema if the table_reference catalog_name or db_schema_name is None. Did you intend to implement this?

Reply:
We will now resolve all possible matches. If >1 match, raise error.